otherlibs/dune-rpc/private/dune_rpc_private.ml

```diff
@@ -123,11 +160,12 @@ module Diagnostic = struct
   end
 
   type t =
-    { targets : string list
+    { targets : Target.t list
     ; message : unit Stdune.Pp.t
     ; loc : Loc.t option
     ; severity : severity option
     ; promotion : Promotion.t list
+    ; directory : string option
```

Comment:
In which case would `directory` be `None`?

Reply:
I briefly touched into this in the doc on this field in `dune_rpc.mli`, but to sum up, this would be `None` if the action doesn't have an associated directory. For example, in `dune_rules/context.ml`, there is an error for a version of opam that is too old.

Reply:
I see, thanks!

I wonder if these shouldn't be disambiguated properly outside of this record and have specific records for the different kinds of diagnostics (build, environment, etc.). IIUC, for environment errors such as an incorrect version opam, it doesn't really make sense to have `promotion`, `loc` and `directory`.

But it's outside of the scope of this PR anyways 🙂

Reply:
Sometimes the environment is set within dune, so we can provide the location.

In general, I prefer to give the client the minimum amount of the information they need to do everything correctly. In this case, I'd prefer if clients did not know anything about the errors except what they need to display to the users. If there's a concrete reason for the client to distinguish a class of errors, we can revisit this.